teps/0076-array-result-types.md

```diff
@@ -0,0 +1,321 @@
+  initially support arrays of arrays)
+    * The TaskRun (or Pipeline Task) would fail
+* What if the type of the result content doesn't match the expected type? e.g. an array provided for a string result
+  or vice versa?
+    * The TaskRun (or Pipeline Task) would fail
+* What if the Task writes to both the string file and the json file for the same param?
+    * The TaskRun (or Pipeline Task) would fail
+
+### Risks and Mitigations
+
+* Changing the type of Task Result values from string to being ArrayOrString might be backwards incompatible
+    * Mitigation: perform a manual test to make sure this isn't backwards incompatible (reconcile a TaskRun with a string
+      result both before and after the change - e.g. create a TaskRun before the change, then restart the controller so
+      it is re-reconciled; revisit this proposal if we can't find a way to make the change and be backward compatible)
+    * Note this will be a change for folks consuming the library, e.g. the CLI, as a result implementing this TEP
+      should include working with the CLI team to pull in the updates and make sure they are not negatively impacted
+
+## Design Details
+
+### Emitting array results
+
+We add support for writing json content to `/tekton/results/resultName.json`, supporting strings and arrays of strings.
+
+For example, say we want the value of the Task foo's result `animals` to be "cat", "dog" and "squirrel":
+
+1. Write the following content to `$(results.animals.jsonPath)`:
+
+   ```json
+   ["cat", "dog", "squirrel"]
+   ```
+
+2. This would be written to the pod termination message as escaped json, for example (with a string example included as well):
+
+   ```
+   message: '[{"key":"animals","value":"[\"cat\", \"dog\", \"squirrel\"]","type":"TaskRunResult"},{"key":"someString","value":"aStringValue","type":"TaskRunResult"}]'
+   ```
+
+3. We would use the same
+   [ArrayOrString type](https://github.com/tektoncd/pipeline/blob/1f5980f8c8a05b106687cfa3e5b3193c213cb66e/pkg/apis/pipeline/v1beta1/param_types.go#L89)
+   (which in TEP-0075 could be expanded to support dictionaries as well) for
+   task results, e.g. for the above example, the TaskRun would contain:
+
+   ```yaml
+     taskResults:
+     - name: someString
+       value: aStringValue
+     - name: animals
+       value:
+       - cat
+       - dog
+       - squirrel
+   ```
+
+### Indexing into arrays with variable replacement
+
+In addition to supporting [JSONPath](https://tools.ietf.org/id/draft-goessner-dispatch-jsonpath-00.html)
+style [star notation](https://github.com/tektoncd/pipeline/issues/2041), we add support for
+[the JSONPath subscript operator](https://tools.ietf.org/id/draft-goessner-dispatch-jsonpath-00.html#name-jsonpath-examples)
+for accessing "child" members of the array by index.
+
+For example:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+...
+spec:
+  params:
+    - environments:
+        - 'staging'
+        - 'qa'
+        - 'prod'
```

Comment:
whoops thats not right, it would look like:

```yaml
  params:
    - name: environments
       value:
        - 'staging'
        - 'qa'
        - 'prod'
```

Reply:
I think it would require `type: array` as well? At least for task params this is required I think.

Reply:
whoops! thanks :D

Reply:
i just realized i dont think i included a single example of how you actually specify a result with that type either XD

Reply:
... and then i forgot to fix it again XD

Reply:
aaaaand the values should be in the pipelinerun not the pipeline .. wheeee.....

Reply:
i think i finally fixed it XD